File: avi/rug/rug.py
import numpy as np

def cube(rug, dim:int, row:int, col:int): 
  for r in range(dim):
    for c in range(dim):
      rug[r+row][c+col] = 1
  return row + dim - 1


def line(rug, row:int):
  (height, width)= rug.shape
  for c in range(width):
     rug[row][c] = 1
  return row


def triangle(rug, dim:int, row:int, col:int, invert:bool=False):
  row_list = list(range(dim))
  for r in row_list:
    # compute the number of columns based on our row
    span = r if not invert else dim - r - 1
    col_list = list(range(col-span, col+span+1))
    for c in col_list:
      rug[row+r][c] = 1
  return row + dim - 1


def diamond(rug, dim:int, row:int, col:int):
  row = triangle(rug, dim, row, col)
  row = triangle(rug, dim, row, col, invert=True)
  return row


def diamond_row(rug, dim:int, row:int):
  (height, width)= rug.shape
  for c in range(dim, width - dim, 2 * dim-1):
    last_row = diamond( rug, dim, row, c)
  return last_row


def triangle_row(rug, dim:int, row:int, invert:bool=False):
  (height, width)= rug.shape
  for c in range(dim, width - dim, 2 * dim-1): # start/stop/step
    last_row = triangle(rug, dim, row, c, invert=invert)
    invert = not invert
  return last_row


def cube_row(rug, dim:int, row:int, col_start=0):
  # compute number of cubes
  (height, width)= rug.shape
  for c in range(col_start, width, dim*2): # start/stop/step
    last_row = cube(rug, dim, row, c)
  return last_row


def printme(rug):
  (height, width) = rug.shape
  for r in range(height):
    for c in range(width):
      print('%' if rug[r][c] else ' ', end="")
    print()


HEIGHT = 30
WIDTH = 30


rug = np.zeros((HEIGHT, WIDTH))
row = line(         rug,    0)
row = line(         rug,    row+1)
row = triangle_row( rug, 3, row+2, False)
row = line(         rug,    row+2)
row = cube_row(     rug, 2, row+1)
row = diamond_row(  rug, 4, row+1)
row = cube_row(     rug, 2, row+1)
row = line(         rug,    row+1)
row = triangle_row( rug, 3, row+2, True)
row = line(         rug,    row+2)
row = line(         rug,    row+1)


printme(rug)

File: avi/rug/rug.java
/*
 * Click `Run` to execute the snippet below!
 */

//import java.io.*;
//import java.util.*;


import java.util.Arrays;

public class Solution
{
  static int _WIDTH = 30;
  static int _HEIGHT = 30;
  static char _SHAPE = '~';

  static char[][] _arr = new char[_WIDTH][_HEIGHT];

  public static void init()
  {
    for (int i = 0; i < _WIDTH; i++)
      Arrays.fill(_arr[i], ' ');
  }

  public static void printme()
  {
    for (int r = 0; r < _HEIGHT; r++)
    {
      for (int c = 0; c < _WIDTH; c++)
        System.out.print( _arr[r][c]);
      System.out.println();
    }
  }

  public static int line(int row)
  {
    for(int c = 0; c < _WIDTH; c++)
      _arr[row][c] = _SHAPE;
    return (row);
  }

  public static int cube(int dim, int row, int col)
  {
    for (int r  = 0; r < dim; r++)
      for (int c = 0; c < dim; c++)
        _arr[r+row][c+col] = _SHAPE;
    return (row + dim - 1);
  }

  public static int checker(int dim, int row)
  {
    int last_row = 0;
    for (int i = 0; i < _WIDTH; i += dim*2)
      last_row = cube(dim, row, i);
    return(last_row);
  }

  public static int triangle(int dim, int row, int col)
  {
    for (int r = 0; r < dim; r++)
    {
      int width = r;
      for (int c = -width; c < width+1; c++)
        _arr[r+row][c+col] = _SHAPE;
    }
    return (row + dim - 1);
   }

   public static int triangle_row(int dim, int row)
  {
    int last_row = 0;
    for (int i = dim; i < _WIDTH; i += dim*2)
      last_row = triangle(dim, row, i);

    return(last_row);
  }

  public static int triangle_inv(int dim, int row, int col)
  {
    for (int r = 0; r < dim; r++)
    {
      int width = dim - (r + 1);
      //==> max width is when r == (dim -1)
      // at  r== dim -1, width = dim - (dim -1) ==> wdith +1
      // then columns go from -1, 0, +1 below
      for (int c = -width; c < width+1; c++)
        _arr[r+row][c+col] = _SHAPE;
    }
    return (row + dim - 1);
  }

  public static int triangle_inv_row(int dim, int row)
  {
    int last_row = 0;
    for (int i = dim; i < _WIDTH; i += dim*2)
      last_row = triangle_inv(dim, row, i);
    return(last_row);
  }

  public static int diamond(int dim, int row, int col)
  {
    int last_row = triangle (dim, row, col);
    last_row = triangle_inv(dim, last_row, col);
    return (last_row);
  }

  public static int diamond_row(int dim, int row)
  {
    int last_row = 0;
    for (int i = dim; i < _WIDTH; i += dim*2)
      last_row = diamond(dim, row, i);

    return(last_row);
  }

  public static void main(String args[])
  {
    init();
    int row = 0;
    row = line(row);
    row = line(row+1);
    row = diamond_row(3, row + 1);
    row = line(row + 1);
    row = checker(2, row + 1);
    row = line(row+1);
    row = triangle_inv_row(3, row+1);
    row = triangle_row(3, row +2);
    row = line(row+1);
    row = checker(2, row + 1);
    row = line(row + 1);
    row = diamond_row(3, row + 1);
    row = line(row + 1);
    row = line(row+1);
   // triangle_inv(5, row, 5);
    //diamond(5, 20, 15);
    //triangle_row(3,20);
    //diamond_row(3, 20);
    printme();
  }
}


// /*
//  * To execute Java, please define "static void main" on a class
//  * named Solution.
//  *
//  * If you need more classes, simply define them inline.
//  */

// class Solution {
//   public static void main(String[] args) {
//     ArrayList<String> strings = new ArrayList<String>();
//     strings.add("Hello, World!");
//     strings.add("Welcome to CoderPad.");
//     strings.add("This pad is running Java " + Runtime.version().feature());

//     for (String string : strings) {
//       System.out.println(string);
//     }
//   }
// }
